File: mcbuild/islands.py
"""More than one island — so alt 1 builds on its own, and alt 2 can come over and build here too.

    python -m mcbuild islands                       # what is known
    python -m mcbuild islands --add main --from out/island_now.litematic --owner Enroniti
    python -m mcbuild islands --where -24200 30000  # which island is this coordinate on

Everything in this project has assumed ONE island, because there was one: `plot.find` reads the
bedrock out of a capture, `profile.yaml` names one `origin_lock`, and `fleet.json` claimed designs
with no idea where they were. That holds right up until a second account has its own island and the
two are visited by the same tooling.

**AN ISLAND IS IDENTIFIED BY ITS BEDROCK, NOT BY WHOSE IT IS.** Every skyblock island has exactly
one bedrock block at its origin, which is already how `plot.py` finds the buildable square. So the
bedrock coordinate IS the island's identity: it is stable, it is discoverable from any capture, and
it does not care who is standing on it. Keying off the account instead would break the moment an
alt walks next door — which is the whole thing being built here.

**OWNERSHIP IS A LABEL, NOT A PERMISSION.** `owner` records whose island it is so a report can read
sensibly; it never decides who may build. Alt 2 building on alt 1's island is the case this exists
to support, and a tool that refused it would be enforcing a rule the server does not have.

The one thing that IS enforced: **a design belongs to the island its cells sit on**, so a claim,
a plot check and a build order are all scoped to that island rather than to the whole world.
"""
from __future__ import annotations

import json
import os
import pathlib

FILE = "islands.json"
# Skyblock islands sit far apart; anything within this of a bedrock is on that island. Generous,
# because the plot is 99 wide and the void between islands is much larger than the error here.
NEAR = 256


def path(schem_dir: str | None = None) -> str:
    from .profile import load as load_profile
    return os.path.join(schem_dir or load_profile()["schem_dir"], FILE)


def load(schem_dir: str | None = None) -> dict:
    p = path(schem_dir)
    if not os.path.exists(p):
        return {"islands": {}}
    try:
        with open(p, encoding="utf-8") as f:
            d = json.load(f) or {}
    except (OSError, ValueError):
        return {"islands": {}}
    d.setdefault("islands", {})
    return d


def save(state: dict, schem_dir: str | None = None) -> str:
    p = path(schem_dir)
    pathlib.Path(p).parent.mkdir(parents=True, exist_ok=True)
    tmp = p + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(state, f, indent=1)
        f.flush()
        os.fsync(f.fileno())
    os.replace(tmp, p)                     # five clients share this, as with fleet.json
    return p


def add(name: str, capture: str, owner: str = "", radius: int | None = None,
        schem_dir: str | None = None) -> dict:
    """Record an island, with its centre DISCOVERED from the capture's bedrock.

    Raises rather than guessing a centre. A registry entry with a made-up origin would send every
    plot check, every siting decision and every claim to the wrong square, and it would look right.
    """
    from . import plot as plot_mod
    pl = plot_mod.find(capture) if radius is None else plot_mod.find(capture, radius)
    state = load(schem_dir)
    state["islands"][name] = {
        "cx": pl.cx, "cz": pl.cz, "radius": pl.radius,
        "owner": owner, "capture": capture,
    }
    save(state, schem_dir)
    return state["islands"][name]


def at(x: int, z: int, schem_dir: str | None = None) -> str | None:
    """Which island this coordinate is on, or None.

    NEAREST BEDROCK WITHIN RANGE, not "inside the plot". A build on the rim, a route between two
    islands and a player in the void between them all need an answer, and "outside every plot" is
    not the same as "nowhere".
    """
    best, bestd = None, None
    for name, isl in load(schem_dir).get("islands", {}).items():
        d = max(abs(x - isl["cx"]), abs(z - isl["cz"]))
        if d <= NEAR and (bestd is None or d < bestd):
            best, bestd = name, d
    return best


def plot_of(name: str, schem_dir: str | None = None):
    """The buildable square of a named island."""
    from .plot import Plot
    isl = load(schem_dir).get("islands", {}).get(name)
    if not isl:
        return None
    return Plot(isl["cx"], isl["cz"], isl.get("radius", 49))


def owner(name: str, schem_dir: str | None = None) -> str:
    return load(schem_dir).get("islands", {}).get(name, {}).get("owner", "")


def island_of_design(design: str, schem_dir: str | None = None) -> str | None:
    """Which island a design's cells sit on, read from its own sidecar origin.

    The design says where it is; nothing has to be told. That is the same contract the sidecars
    have carried since the beginning - no sidecar, no world position.
    """
    from . import scan as scan_mod
    try:
        sc = scan_mod.load(design)
    except Exception:                                            # noqa: BLE001
        return None
    ox, _, oz = sc.origin
    sy, sz, sx = sc.model.ids.shape
    return at(ox + sx // 2, oz + sz // 2, schem_dir)


def report(schem_dir: str | None = None) -> str:
    state = load(schem_dir)
    isls = state.get("islands", {})
    if not isls:
        return ("no islands recorded yet.\n"
                "  python -m mcbuild islands --add main --from out/island_now.litematic "
                "--owner <account>\n"
                "  (the centre is DISCOVERED from the capture's bedrock, never typed)")
    lines = [f"{len(isls)} island(s):"]
    for name, i in sorted(isls.items()):
        r = i.get("radius", 49)
        lines.append(f"  {name:12s} bedrock {i['cx']} {i['cz']}  "
                     f"X {i['cx'] - r}..{i['cx'] + r} Z {i['cz'] - r}..{i['cz'] + r}"
                     + (f"  owner {i['owner']}" if i.get("owner") else ""))
    lines.append("  ownership is a LABEL, not a permission - any alt may build on any of them")
    return "\n".join(lines)

File: chunkscan/src/client/java/dev/jack/chunkscan/Islands.java
package dev.jack.chunkscan;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.minecraft.client.Minecraft;
import net.minecraft.core.BlockPos;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Which island you are standing on — so alt 2 can walk onto alt 1's island and build there.
 *
 * <p>Everything in this mod has assumed one island, because there was one. {@link Plot} carried a
 * single square baked in at export time; {@link Fleet} claimed designs with no idea where they
 * were. Both hold right up until a second account has its own island and the same tooling visits
 * it.
 *
 * <p><b>AN ISLAND IS IDENTIFIED BY ITS BEDROCK, NOT BY WHOSE IT IS.</b> Every skyblock island has
 * exactly one bedrock at its origin, which is already how the plot is found. So the bedrock
 * coordinate IS the identity: stable, discoverable, and indifferent to who is standing on it.
 * Keying off the account would break the moment an alt walks next door, which is the entire case
 * this exists to support.
 *
 * <p><b>OWNERSHIP IS A LABEL, NOT A PERMISSION.</b> It makes a report read sensibly and it never
 * decides who may build — a tool that refused alt 2 on alt 1's island would be enforcing a rule
 * the server does not have.
 *
 * <p>Read from {@code islands.json}, written by {@code python -m mcbuild islands --add}. When the
 * file is absent everything falls back to the single baked-in plot, so a setup that has never
 * heard of a second island behaves exactly as it did.
 */
final class Islands {
	static final String FILE = "islands.json";
	/** Islands sit far apart; the plot is 99 wide and the void between them is much larger. */
	static final int NEAR = 256;

    record Bounds(int minX, int minZ, int maxXExclusive, int maxZExclusive) {
        Bounds {
            if (minX >= maxXExclusive || minZ >= maxZExclusive) throw new IllegalArgumentException("Empty/inverted island bounds");
        }
        boolean contains(int x, int z) {
            return x >= minX && x < maxXExclusive && z >= minZ && z < maxZExclusive;
        }
        int over(int x, int z) {
            long dx = Math.max(0L, Math.max((long) minX - x, (long) x - maxXExclusive + 1));
            long dz = Math.max(0L, Math.max((long) minZ - z, (long) z - maxZExclusive + 1));
            return (int) Math.min(Integer.MAX_VALUE, Math.max(dx, dz));
        }
    }

    record Island(String name, int cx, int cz, int radius, String owner, Bounds bounds, String site) {
        boolean contains(int x, int z) { return bounds.contains(x, z); }
        int over(int x, int z) { return bounds.over(x, z); }
    }

    private static Island parse(String name, JsonObject o) {
        int cx = o.get("cx").getAsInt(), cz = o.get("cz").getAsInt();
        int radius = o.has("radius") ? o.get("radius").getAsInt() : 49;
        if (radius < 0) throw new IllegalArgumentException("Negative island radius");
        Bounds bounds;
        if (o.has("bounds")) {
            JsonObject b = o.getAsJsonObject("bounds");
            bounds = new Bounds(b.get("min_x").getAsInt(), b.get("min_z").getAsInt(),
                b.get("max_x_exclusive").getAsInt(), b.get("max_z_exclusive").getAsInt());
        } else bounds = new Bounds(Math.subtractExact(cx, radius), Math.subtractExact(cz, radius),
            Math.addExact(Math.addExact(cx, radius), 1), Math.addExact(Math.addExact(cz, radius), 1));
        if (!bounds.contains(cx, cz)) throw new IllegalArgumentException("Bedrock outside island bounds");
        String site = o.has("site") ? o.get("site").getAsString() : name;
        if (site.isBlank()) throw new IllegalArgumentException("Empty island site");
        return new Island(name, cx, cz, radius, o.has("owner") ? o.get("owner").getAsString() : "", bounds, site);
    }

	// KEYED BY DIRECTORY, not just by time. Cached on the clock alone, a second schematics folder
	// silently gets the first one's islands — which is benign today (there is one folder) and is
	// exactly the latent bug that bites the day there are two. Its own test found this.
	private static final Map<String, Map<String, Island>> cache = new LinkedHashMap<>();
	private static final Map<String, Long> cachedAt = new LinkedHashMap<>();

	private Islands() {}

	static synchronized Map<String, Island> all(Path dir) {
        dir = ActiveBuild.siteInputs(dir);
		// Re-read now and then rather than once: the registry is written by the Python side while
		// the game is running, and a client that cached it at login would never see a new island.
		String key = dir.toString();
		Long when = cachedAt.get(key);
		if (when != null && System.currentTimeMillis() - when < 30_000) return cache.get(key);
		Map<String, Island> out = new LinkedHashMap<>();
		Path f = dir.resolve(FILE);
		if (Files.exists(f)) {
			try {
				JsonObject root = JsonParser.parseString(Files.readString(f, StandardCharsets.UTF_8))
					.getAsJsonObject();
				JsonObject isls = root.getAsJsonObject("islands");
				if (isls != null) {
					for (var e : isls.entrySet()) {
						JsonObject o = e.getValue().getAsJsonObject();
                        out.put(e.getKey(), parse(e.getKey(), o));
					}
				}
			} catch (Exception e) {
				ChunkScanClient.LOG.warn("islands.json unreadable: {}", e.toString());
                out.clear(); // Never retain a partially parsed authorization registry.
			}
		}
		cache.put(key, out);
		cachedAt.put(key, System.currentTimeMillis());
		return out;
	}

	/**
	 * The island nearest this coordinate, or null.
	 *
	 * <p>NEAREST BEDROCK WITHIN RANGE, not "inside the plot": a build on the rim, a flight between
	 * two islands and a player hanging in the void between them all need an answer, and "outside
	 * every plot" is not the same as "nowhere".
	 */
	static Island at(Path dir, int x, int z) {
		Island best = null;
		int bestD = Integer.MAX_VALUE;
		for (Island i : all(dir).values()) {
            if (i.contains(x, z)) return i;
			int d = Math.max(Math.abs(x - i.cx()), Math.abs(z - i.cz()));
			if (d <= NEAR && d < bestD) {
				best = i;
				bestD = d;
			}
		}
		return best;
	}

	/** The island the player is standing on, or null. */
	static Island here(Minecraft mc, Path dir) {
		if (mc.player == null) return null;
		BlockPos p = mc.player.blockPosition();
		return at(dir, p.getX(), p.getZ());
	}

    /** Known registries authorize only their exact plot union. Missing registries retain legacy Plot behavior. */
    static boolean outside(Path dir, int x, int z) {
        Map<String, Island> islands = all(dir);
        if (islands.isEmpty()) return Files.exists(ActiveBuild.siteInputs(dir).resolve(FILE)) || Plot.outside(x, z);
        return islands.values().stream().noneMatch(i -> i.contains(x, z));
    }

    static boolean storageOnSite(Path dir, Island source, int x, int z) {
        return all(dir).values().stream().anyMatch(i -> i.site().equals(source.site()) && i.contains(x, z));
    }

	static int over(Path dir, int x, int z) {
		Island i = at(dir, x, z);
		if (i == null) return all(dir).isEmpty() ? Plot.over(x, z) : 0;
		return i.over(x, z);
	}

	static String describe(Minecraft mc, Path dir) {
		Map<String, Island> all = all(dir);
		if (all.isEmpty()) {
			return Files.exists(ActiveBuild.siteInputs(dir).resolve(FILE)) ? "island registry empty or invalid — automatic placement disabled" : "no island registry — falling back to the single plot: " + Plot.describe()
				+ "\n  python -m mcbuild islands --add <name> --from <capture> to record more";
		}
		Island h = here(mc, dir);
		StringBuilder b = new StringBuilder(all.size() + " island(s) known");
		if (h != null) {
			b.append("\n  you are on ").append(h.name())
				.append(h.owner().isBlank() ? "" : " (" + h.owner() + "'s)")
				.append("  X ").append(h.bounds().minX()).append("..").append(h.bounds().maxXExclusive() - 1)
				.append(" Z ").append(h.bounds().minZ()).append("..").append(h.bounds().maxZExclusive() - 1);
			if (mc.player != null) {
				int o = h.over(mc.player.getBlockX(), mc.player.getBlockZ());
				b.append(o > 0 ? "  — you are " + o + " OUTSIDE it" : "  — you are inside it");
			}
		} else {
			b.append("\n  you are not on any island the registry knows");
		}
		for (Island i : all.values()) {
			if (h != null && i.name().equals(h.name())) continue;
			b.append("\n  ").append(i.name()).append("  bedrock ").append(i.cx()).append(" ")
				.append(i.cz()).append(i.owner().isBlank() ? "" : "  owner " + i.owner());
		}
		b.append("\n  ownership is a LABEL: any alt may build on any of them");
		return b.toString();
	}
}
